Modify upgradeAvailable to adhere to supported upgrade path

jordojordo opened the pull request:
Fix #578 

This changes the `upgradeAvailable` property to only return supported upgrades based on the [documented upgrade path](https://docs.kubewarden.io/next/operator-manual/upgrade-path#upgrade-paths).


https://github.com/rancher/kubewarden-ui/assets/40806497/2ad046e3-ee47-4be9-b6a5-cbab3ad54221

aalves08 reviewed (commented):
Right in the end of the video, shouldn't the blue "box" show the `appVersion` `v1.9.0`? That's the latest installed `version` `2.0.5`... Is that correct?

![Screenshot 2024-01-04 at 18 48 56](https://github.com/rancher/kubewarden-ui/assets/97888974/09924466-34ee-4a65-91e1-0ff61f788314)

Looking at the Kubewarden documentation and the behaviour shown, I agree with the upgrade path

kravciak commented:
Looks good but I was not able to test all paths because of https://github.com/rancher/kubewarden-ui/issues/580.

- If current version is 1.1.0 should I upgrade to 1.1.2 before going to 1.2.1? 
![Screenshot from 2024-01-05 12-03-37](https://github.com/rancher/kubewarden-ui/assets/538604/eca732d5-9165-44e6-8ce4-e2f1f8b435f4)

- Do we take APP versions into account or only CHART version? For example cases where (or was this just wrong release?)
   - Chart (MINOR) changes but APP does not (1.2.8 -> 1.3.0)
   - Chart (MINOR) does not change but APP does (1.2.7 -> 1.2.8)
```
NAME                 	CHART VERSION	APP VERSION
kubewarden/kubewarden-controller	1.5.1        	v1.6.0
kubewarden/kubewarden-controller	1.5.0        	v1.6.0
kubewarden/kubewarden-controller	1.3.1        	v1.4.0
kubewarden/kubewarden-controller	1.3.0        	v1.4.0
kubewarden/kubewarden-controller	1.2.8        	v1.4.0
kubewarden/kubewarden-controller	1.2.7        	v1.3.0
kubewarden/kubewarden-controller	1.2.6        	v1.3.0
```

@viccuad might know?

jordojordo commented:
I've updated this with clear indication on what can be updated. The upgrade button now includes the Chart `version` and `appVersion`, the computation behind the upgrade path now checks the `chart.version` rather than the `chart.appVersion` for compatibility.

![ver](https://github.com/rancher/kubewarden-ui/assets/40806497/87f29977-3ba4-4f34-a926-a0f24c23d14e)

aalves08 reviewed (approved):
LGTM

jordojordo commented:
I updated this to take the user preference for pre-release Helm chart versions into account when presenting upgrades, and added unit tests for this case as well.


https://github.com/rancher/kubewarden-ui/assets/40806497/6c4b9f29-aa79-4a0d-a8e4-67e6a83505fd

aalves08 reviewed (approved):
LGTM

jordojordo commented:
@kravciak I updated this to follow the flow mentioned by @viccuad - it will now check for updates to `appVersion` initially, and then the chart `version`. Also added a test for this case.

kravciak commented:
But now we skip upgrade to latest chart versions of current app.

For example I installed kubewarden-controller `chart v1.5.0`. Upgrade path skips `chart 1.5.3`:
 - expected: `app 1.6.0 (chart 1.5.0)` -> `app 1.6.0 (chart 1.5.3)` -> `app 1.7.0 (chart 1.6.2)` -> `app 1.8.0`
 - current:  `app 1.6.0 (chart 1.5.0)` -> `app 1.7.0 (chart 1.6.2)` -> `app 1.7.0` -> `app 1.8.0` -> `app 1.9.0`

```
kubewarden/kubewarden-controller	1.6.2        	v1.7.0     	A Helm chart for deploying the Kubewarden stack
kubewarden/kubewarden-controller	1.6.1        	v1.7.0     	A Helm chart for deploying the Kubewarden stack
kubewarden/kubewarden-controller	1.6.0        	v1.7.0     	A Helm chart for deploying the Kubewarden stack
kubewarden/kubewarden-controller	1.5.3        	v1.6.0     	A Helm chart for deploying the Kubewarden stack
kubewarden/kubewarden-controller	1.5.2        	v1.6.0     	A Helm chart for deploying the Kubewarden stack
kubewarden/kubewarden-controller	1.5.1        	v1.6.0     	A Helm chart for deploying the Kubewarden stack
kubewarden/kubewarden-controller	1.5.0        	v1.6.0     	A Helm chart for deploying the Kubewarden stack
```

viccuad commented:
@kravciak I think that's correct, the supported upgrade paths are all taking appVersion as the semver to check against.

There aren't any patch `appVersion` between `1.6.0` to `1.7.0`, or `1.7.0` to `1.8.0`, or `1.8.0` to `1.9.0`.

Hence it should be possible to upgrade from chart `appVersion 1.6.0 (version 1.5.0)` to `appversion 1.7.0 (version 1.6.2)`. If it weren't, we should have released a chart `appVersion 1.6.1 (version whatever, potentially `1.5.1`)`.
